Lombiq.ChartJs.Samples/Controllers/SampleController.cs

```diff
@@ -0,0 +1,214 @@
+                {
+                    Duration = 0,
+                },
+                Layout = new
+                {
+                    AutoPadding = false,
+                    Padding = 0,
+                    DevicePixelRatio = 1,
+                },
+            },
+        });
+    // NEXT STATION: Go to Views/Balance.cshtml.
+
+    // Generates a view(Views/History.cshtml) with line chart using <chart> tag helper with two lines to display the monthly amount
+    // of incomes and expenses.
+    // Both of them can be filtered with a tag assigned to it. The tag comes from OC taxonomy.
+    // It uses data from Income and Expense content types stored in db as dataset to it.
+    // Lombiq.ChartJs.Samples/Sample/History
+    public async Task<IActionResult> History(string incomeTag = null, string expenseTag = null)
+    {
+        var incomeTagsFilter = await GetItemIdsByTermIdAsync(ContentItemIds.IncomeTagsTaxonomy, incomeTag);
+        var expenseTagsFilter = await GetItemIdsByTermIdAsync(ContentItemIds.ExpenseTagsTaxonomy, expenseTag);
+
+        var transactions = (await _session.QueryIndex<NumericFieldIndex>(
+            index =>
+                index.Published &&
+                index.Latest &&
+                (index.ContentType == ContentTypes.Income || index.ContentType == ContentTypes.Expense) &&
+                (index.ContentPart == nameof(IncomePart) || index.ContentPart == nameof(ExpensePart)) &&
+                index.ContentField == nameof(TransactionPart.Amount) &&
+                index.Numeric != null)
+            .ListAsync())
+            .Join(
+                await _session.QueryIndex<DateFieldIndex>(
+                index =>
+                    index.Published &&
+                    index.Latest &&
+                    (index.ContentType == ContentTypes.Income || index.ContentType == ContentTypes.Expense) &&
+                    (index.ContentPart == nameof(IncomePart) || index.ContentPart == nameof(ExpensePart)) &&
+                    index.ContentField == nameof(TransactionPart.Date) &&
+                    index.Date != null)
+                .ListAsync(),
+                index => index.ContentItemId,
+                index => index.ContentItemId,
+                (numericIndex, dateIndex) =>
+                    new
+                    {
+                        numericIndex.ContentItemId,
+                        numericIndex.ContentType,
+                        Date = dateIndex.Date.Value,
+                        Amount = numericIndex.Numeric.Value,
+                    })
+            .Where(
+                transaction =>
+                    (transaction.ContentType == ContentTypes.Income &&
+                        (incomeTagsFilter == null || incomeTagsFilter.Contains(transaction.ContentItemId))) ||
+                    (transaction.ContentType == ContentTypes.Expense &&
+                        (expenseTagsFilter == null || expenseTagsFilter.Contains(transaction.ContentItemId))))
```

Comment:
Why are these checks done offline?
Can't you use `(incomeTagsFilter == null || index.ContentItemId.IsIn(incomeTagsFilter))` and the like in the index query above?

Reply:
What you mean "checking offline"?

If "checking offline" means checking inside predicate of _session.QueryIndex, then the predicate is a Linq expression, and I can use only constants inside. incomeTagsFilter and expenseTagsFilter are not constants.

Reply:
"Offline" from the web server's perspective. In other words after the query results have been downloaded from the database. Online would be filtering as part of the query so it's evaluated by the database engine and only the matching items sent back to the web server.

> the predicate is a Linq expression, and I can use only constants inside.

That's not true, you can use mutable variables in Linq expressions. You can see a similar call that filters with a `List` using the `value.IsIn(enumerable)` extension method [here](https://github.com/Lombiq/Helpful-Libraries/blob/4f8808b3bd7b35d1ebb1914a7b286b437e08fd21/Lombiq.HelpfulLibraries.Tests/UnitTests/Services/ManualConnectingIndexServiceTests.cs#L36).

Reply:
Ok, I'll check this. Thank you.